hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/LowRedundancyBlocks.java

```diff
@@ -369,7 +369,7 @@ synchronized boolean remove(BlockInfo block,
    * @return true if the block was found and removed from one of the priority
    *         queues
    */
-  boolean remove(BlockInfo block, int priLevel) {
+  synchronized boolean remove(BlockInfo block, int priLevel) {
```

Comment:
should `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` be made synchronized instead. Its other callers are synchronized methods:
1. https://github.com/apache/hadoop/blob/01027e52a9789eb5b386729f52b7bb9e52fa5352/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/LowRedundancyBlocks.java#L345
2. https://github.com/apache/hadoop/blob/01027e52a9789eb5b386729f52b7bb9e52fa5352/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/LowRedundancyBlocks.java#L468

Reply:
This will just ensure only one thread getting into the remove method. But what if there is one thread trying to remove, other trying to add or size(). I feel following should help:
```
private Object syncOnBlockInfoSet(int priority, BlockInfo blockInfo, String operation) {
    LightWeightLinkedSet<BlockInfo> blockInfos =priorityQueues.get(priority); 
    synchronized (blockInfos) {
      if("size".equalsIgnoreCase(operation)) {
        return blockInfos.size();
      }
      if("remove".equalsIgnoreCase(operation)) {
        return blockInfos.remove(blockInfo);
      }
      //implement other required methods.
      else {
        return null;
      }
    }
  }
```

Reply:
> should `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` be made synchronized instead. Its other callers are synchronized methods:

Sorry I don't quite understand.  Since the callers are already synchronized, why is it necessary to made ```boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)``` synchronized?

Reply:
> This will just ensure only one thread getting into the remove method. But what if there is one thread trying to remove, other trying to add or size(). 

add() and size() are already synchronized in the current code.

1. https://github.com/apache/hadoop/blob/01027e52a9789eb5b386729f52b7bb9e52fa5352/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/LowRedundancyBlocks.java#L290
2. https://github.com/apache/hadoop/blob/01027e52a9789eb5b386729f52b7bb9e52fa5352/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/LowRedundancyBlocks.java#L125

Reply:
> > should `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` be made synchronized instead. Its other callers are synchronized methods:
> 
> Sorry I don't quite understand. Since the callers are already synchronized, why is it necessary to made `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` synchronized?

Since  `boolean remove(BlockInfo block, int priLevel)` forwards the call to `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)`, I am suggesting that lets make `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` synchronized instead.
Callers of `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` are synchronized means that we can have `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` as synchronized without any perf-loss.

Reply:
> > This will just ensure only one thread getting into the remove method. But what if there is one thread trying to remove, other trying to add or size().
> 
> add() and size() are already synchronized in the current code.
> 
> 1. https://github.com/apache/hadoop/blob/01027e52a9789eb5b386729f52b7bb9e52fa5352/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/LowRedundancyBlocks.java#L290
> 2. https://github.com/apache/hadoop/blob/01027e52a9789eb5b386729f52b7bb9e52fa5352/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/LowRedundancyBlocks.java#L125

these method being synchronized means that only one thread can enter into the synchronized method, but doesn't make the object its working on synchronized.  There could be one thread calling size which leads to `priorityQueues.get(i).size()`, and other calling add which leads to `priorityQueues.get(priLevel).add(blockInfo)` simultaneously.
What probable issue is:
for adding element https://github.com/apache/hadoop/blob/ccdb978603696a91c4828a66aad27f585543b376/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/util/LightWeightLinkedSet.java#L87-L125 is the code. let say thread goes till https://github.com/apache/hadoop/blob/ccdb978603696a91c4828a66aad27f585543b376/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/util/LightWeightLinkedSet.java#L101 and then context switch happens and the thread doesn't get CPU in future for some time. Now size has been incremented but addition has not happend. the thread calling size() gets the CPU, it will get size which is more than what exactly is in the set. Thats why feel that priorityQueues is still not thread-safe and hence suggesting the change.
Feel free to disagree.

Thanks.

Reply:
Thanks for your suggestion, I'll make both `boolean remove(BlockInfo block, int priLevel, int oldExpectedReplicas)` and `boolean remove(BlockInfo block, int priLevel)` synchronized to ensure correctness.

However, I think you misunderstood the semantics of `synchronized` a method.  Refer to [java doc](https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html):
>  When one thread is executing a synchronized method for an object, all other threads that invoke synchronized methods for the same object block (suspend execution) until the first thread is done with the object.

Reply:
@saxenapranav Thanks for your pretty review comment. I am not sure if get your points totally. IMO, this improvement is safe and self-contained, because `synchronized` is reentrant and exclusive. So I am confused if it could involve other consistency issues. 
I would like to give my +1 if you were worried about perf-loss only for `synchronized-synchronized`, for this case I think it could be acceptable, anyway totally agree that both changes about performance we should given the benchmark comparison.
Thanks @zhangshuyan0 and @saxenapranav , Please feel free to correct me if something wrong.

Reply:
Agreeing with both @Hexiaoqiao  @zhangshuyan0.
Thanks @zhangshuyan0  for taking suggestion.